File: lib/web_ui/lib/src/engine/skwasm/skwasm_impl/src/path.dart
import 'dart:ffi';
import 'dart:typed_data';

import 'package:ui/src/engine.dart' as engine;
import 'package:ui/ui.dart' as ui;
import 'raw/raw_memory.dart';
import 'raw/raw_path.dart';

enum PathDirection {
  clockwise,
  counterClockwise,
}

enum PathArcSize {
  small,
  large,
}

class SkwasmPath implements ui.Path {
  factory SkwasmPath() {
    return SkwasmPath._fromHandle(pathCreate());
  }

  factory SkwasmPath.from(SkwasmPath source) {
    return SkwasmPath._fromHandle(pathCopy(source._handle));
  }

  SkwasmPath._fromHandle(this._handle);
  final PathHandle _handle;

  PathHandle get handle => _handle;

  void delete() {
    pathDestroy(_handle);
  }

  @override
  ui.PathFillType get fillType => ui.PathFillType.values[pathGetFillType(_handle)];

  @override
  set fillType(ui.PathFillType fillType) => pathSetFillType(_handle, fillType.index);

  @override
  void moveTo(double x, double y) => pathMoveTo(_handle, x, y);

  @override
  void relativeMoveTo(double x, double y) => pathRelativeMoveTo(_handle, x, y);

  @override
  void lineTo(double x, double y) => pathLineTo(_handle, x, y);

  @override
  void relativeLineTo(double x, double y) => pathRelativeMoveTo(_handle, x, y);

  @override
  void quadraticBezierTo(double x1, double y1, double x2, double y2) =>
    pathQuadraticBezierTo(_handle, x1, y1, x2, y2);

  @override
  void relativeQuadraticBezierTo(double x1, double y1, double x2, double y2) =>
    pathRelativeQuadraticBezierTo(_handle, x1, y1, x2, y2);

  @override
  void cubicTo(
    double x1,
    double y1,
    double x2,
    double y2,
    double x3,
    double y3) =>
    pathCubicTo(_handle, x1, y1, x2, y2, x3, y3);

  @override
  void relativeCubicTo(
      double x1,
      double y1,
      double x2,
      double y2,
      double x3,
      double y3) =>
    pathRelativeCubicTo(_handle, x1, y1, x2, y2, x3, y3);

  @override
  void conicTo(double x1, double y1, double x2, double y2, double w) =>
    pathConicTo(_handle, x1, y1, x2, y2, w);

  @override
  void relativeConicTo(double x1, double y1, double x2, double y2, double w) =>
    pathRelativeConicTo(_handle, x1, y1, x2, y2, w);

  @override
  void arcTo(
      ui.Rect rect, double startAngle, double sweepAngle, bool forceMoveTo) {
    withStackScope((StackScope s) {
      pathArcToOval(
          _handle,
          s.convertRect(rect),
          ui.toDegrees(startAngle),
          ui.toDegrees(sweepAngle),
          forceMoveTo
      );
    });
  }

  @override
  void arcToPoint(
    ui.Offset arcEnd, {
    ui.Radius radius = ui.Radius.zero,
    double rotation = 0.0,
    bool largeArc = false,
    bool clockwise = true,
  }) {
    final PathArcSize arcSize =
        largeArc ? PathArcSize.large : PathArcSize.small;
    final PathDirection pathDirection =
        clockwise ? PathDirection.clockwise : PathDirection.counterClockwise;
    pathArcToRotated(
        _handle,
        radius.x,
        radius.y,
        ui.toDegrees(rotation),
        arcSize.index,
        pathDirection.index,
        arcEnd.dx,
        arcEnd.dy
    );
  }

  @override
  void relativeArcToPoint(
    ui.Offset arcEndDelta, {
    ui.Radius radius = ui.Radius.zero,
    double rotation = 0.0,
    bool largeArc = false,
    bool clockwise = true,
  }) {
    final PathArcSize arcSize =
        largeArc ? PathArcSize.large : PathArcSize.small;
    final PathDirection pathDirection =
        clockwise ? PathDirection.clockwise : PathDirection.counterClockwise;
    pathRelativeArcToRotated(
        _handle,
        radius.x,
        radius.y,
        ui.toDegrees(rotation),
        arcSize.index,
        pathDirection.index,
        arcEndDelta.dx,
        arcEndDelta.dy
    );
  }

  @override
  void addRect(ui.Rect rect) {
    withStackScope((StackScope s) {
      pathAddRect(_handle, s.convertRect(rect));
    });
  }

  @override
  void addOval(ui.Rect rect) {
    withStackScope((StackScope s) {
      pathAddOval(_handle, s.convertRect(rect));
    });
  }

  @override
  void addArc(ui.Rect rect, double startAngle, double sweepAngle) {
    withStackScope((StackScope s) {
      pathAddArc(
        _handle,
        s.convertRect(rect),
        ui.toDegrees(startAngle),
        ui.toDegrees(sweepAngle)
      );
    });
  }

  @override
  void addPolygon(List<ui.Offset> points, bool close) {
    withStackScope((StackScope s) {
      pathAddPolygon(_handle, s.convertPointArray(points), points.length, close);
    });
  }

  @override
  void addRRect(ui.RRect rrect) {
    withStackScope((StackScope s) {
      pathAddRRect(_handle, s.convertRRect(rrect));
    });
  }

  @override
  void addPath(ui.Path path, ui.Offset offset, {Float64List? matrix4}) {
    _addPath(path, offset, false, matrix4: matrix4);
  }

  @override
  void extendWithPath(ui.Path path, ui.Offset offset, {Float64List? matrix4}) {
    _addPath(path, offset, true, matrix4: matrix4);
  }

  void _addPath(ui.Path path, ui.Offset offset, bool extend, {Float64List? matrix4}) {
    assert(path is SkwasmPath);
    withStackScope((StackScope s) {
      final Pointer<Float> convertedMatrix =
          s.convertMatrix4toSkMatrix(matrix4 ?? engine.Matrix4.identity().toFloat64());
      convertedMatrix[2] += offset.dx;
      convertedMatrix[5] += offset.dy;
      pathAddPath(_handle, (path as SkwasmPath)._handle, convertedMatrix, extend);
    });
  }

  @override
  void close() => pathClose(_handle);

  @override
  void reset() => pathReset(_handle);

  @override
  bool contains(ui.Offset point) => pathContains(_handle, point.dx, point.dy);

  @override
  ui.Path shift(ui.Offset offset) => 
    transform(engine.Matrix4.translationValues(offset.dx, offset.dy, 0.0).toFloat64());

  @override
  ui.Path transform(Float64List matrix4) {
    return withStackScope((StackScope s) {
      final PathHandle newPathHandle = pathCopy(_handle);
      pathTransform(newPathHandle, s.convertMatrix4toSkMatrix(matrix4));
      return SkwasmPath._fromHandle(newPathHandle);
    });
  }

  @override
  ui.Rect getBounds() {
    return withStackScope((StackScope s) {
      final Pointer<Float> rectBuffer = s.allocFloatArray(4);
      pathGetBounds(_handle, rectBuffer);
      return ui.Rect.fromLTRB(
        rectBuffer[0], 
        rectBuffer[1], 
        rectBuffer[2], 
        rectBuffer[3]
      );
    });
  }

  static SkwasmPath combine(
    ui.PathOperation operation, 
    SkwasmPath path1, 
    SkwasmPath path2) =>
    SkwasmPath._fromHandle(pathCombine(
        operation.index, path1._handle, path2._handle));
  
  @override
  ui.PathMetrics computeMetrics({bool forceClosed = false}) {
    // TODO: implement computeMetrics
    throw UnimplementedError();
  }
}

File: lib/web_ui/skwasm/path.cpp
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "export.h"
#include "helpers.h"
#include "render_strategy.h"

using namespace Skwasm;

SKWASM_EXPORT Path* path_create() {
  return new Path();
}

SKWASM_EXPORT void path_dispose(Path* path) {
  delete path;
}

SKWASM_EXPORT Path* path_copy(Path* path) {
  return new Path(*path);
}

SKWASM_EXPORT void path_setFillType(Path* path, PathFillType fillType) {
  path->setFillType(fillType);
}

SKWASM_EXPORT PathFillType path_getFillType(Path* path) {
  return path->getFillType();
}

SKWASM_EXPORT void path_moveTo(Path* path, Scalar x, Scalar y) {
  path->moveTo(x, y);
}

SKWASM_EXPORT void path_relativeMoveTo(Path* path, Scalar x, Scalar y) {
  path->rMoveTo(x, y);
}

SKWASM_EXPORT void path_lineTo(Path* path, Scalar x, Scalar y) {
  path->lineTo(x, y);
}

SKWASM_EXPORT void path_relativeLineTo(Path* path, Scalar x, Scalar y) {
  path->rLineTo(x, y);
}

SKWASM_EXPORT void path_quadraticBezierTo(Path* path,
                                          Scalar x1,
                                          Scalar y1,
                                          Scalar x2,
                                          Scalar y2) {
  path->quadTo(x1, y1, x2, y2);
}

SKWASM_EXPORT void path_relativeQuadraticBezierTo(Path* path,
                                                  Scalar x1,
                                                  Scalar y1,
                                                  Scalar x2,
                                                  Scalar y2) {
  path->rQuadTo(x1, y1, x2, y2);
}

SKWASM_EXPORT void path_cubicTo(Path* path,
                                Scalar x1,
                                Scalar y1,
                                Scalar x2,
                                Scalar y2,
                                Scalar x3,
                                Scalar y3) {
  path->cubicTo(x1, y1, x2, y2, x3, y3);
}

SKWASM_EXPORT void path_relativeCubicTo(Path* path,
                                        Scalar x1,
                                        Scalar y1,
                                        Scalar x2,
                                        Scalar y2,
                                        Scalar x3,
                                        Scalar y3) {
  path->rCubicTo(x1, y1, x2, y2, x3, y3);
}

SKWASM_EXPORT void path_conicTo(Path* path,
                                Scalar x1,
                                Scalar y1,
                                Scalar x2,
                                Scalar y2,
                                Scalar w) {
  path->conicTo(x1, y1, x2, y2, w);
}

SKWASM_EXPORT void path_relativeConicTo(Path* path,
                                        Scalar x1,
                                        Scalar y1,
                                        Scalar x2,
                                        Scalar y2,
                                        Scalar w) {
  path->rConicTo(x1, y1, x2, y2, w);
}

SKWASM_EXPORT void path_arcToOval(Path* path,
                                  const Rect* rect,
                                  Scalar startAngle,
                                  Scalar sweepAngle,
                                  bool forceMoveTo) {
  path->arcTo(*rect, startAngle, sweepAngle, forceMoveTo);
}

SKWASM_EXPORT void path_arcToRotated(Path* path,
                                     Scalar rx,
                                     Scalar ry,
                                     Scalar xAxisRotate,
                                     Path::ArcSize arcSize,
                                     PathDirection pathDirection,
                                     Scalar x,
                                     Scalar y) {
  path->arcTo(rx, ry, xAxisRotate, arcSize, pathDirection, x, y);
}

SKWASM_EXPORT void path_relativeArcToRotated(Path* path,
                                             Scalar rx,
                                             Scalar ry,
                                             Scalar xAxisRotate,
                                             Path::ArcSize arcSize,
                                             PathDirection pathDirection,
                                             Scalar x,
                                             Scalar y) {
  path->rArcTo(rx, ry, xAxisRotate, arcSize, pathDirection, x, y);
}

SKWASM_EXPORT void path_addRect(Path* path, const Rect* rect) {
  path->addRect(*rect);
}

SKWASM_EXPORT void path_addOval(Path* path, const Rect* oval) {
  path->addOval(*oval, PathDirection::kCW, 1);
}

SKWASM_EXPORT void path_addArc(Path* path,
                               const Rect* oval,
                               Scalar startAngle,
                               Scalar sweepAngle) {
  path->addArc(*oval, startAngle, sweepAngle);
}

SKWASM_EXPORT void path_addPolygon(Path* path,
                                   const Point* points,
                                   int count,
                                   bool close) {
  path->addPoly(points, count, close);
}

SKWASM_EXPORT void path_addRRect(Path* path, const Scalar* rrectValues) {
  path->addRRect(createRRect(rrectValues), PathDirection::kCW);
}

SKWASM_EXPORT void path_addPath(Path* path,
                                const Path* other,
                                const Scalar* matrix33,
                                Path::AddPathMode extendPath) {
  path->addPath(*other, createMatrix(matrix33), extendPath);
}

SKWASM_EXPORT void path_close(Path* path) {
  path->close();
}

SKWASM_EXPORT void path_reset(Path* path) {
  path->reset();
}

SKWASM_EXPORT bool path_contains(Path* path, Scalar x, Scalar y) {
  return path->contains(x, y);
}

SKWASM_EXPORT void path_transform(Path* path, const Scalar* matrix33) {
  path->transform(createMatrix(matrix33));
}

SKWASM_EXPORT void path_getBounds(Path* path, Rect* rect) {
  *rect = path->getBounds();
}

SKWASM_EXPORT Path* path_combine(PathOp operation,
                                 const Path* path1,
                                 const Path* path2) {
  Path* output = new Path();
  if (Op(*path1, *path2, operation, output)) {
    output->setFillType(path1->getFillType());
    return output;
  } else {
    delete output;
    return nullptr;
  }
}
